lib/broadway/topology/batcher_stage.ex

```diff
@@ -110,36 +110,112 @@ defmodule Broadway.Topology.BatcherStage do
   defp handle_events_per_batch_key([event | _] = events, acc, state) do
     %{partition_by: partition_by} = state
     batch_key = batch_key(event, partition_by)
-    {current, pending_count, timer} = init_or_get_batch(batch_key, state)
+    {current, pending_count, batch_splitter, timer} = init_or_get_batch(batch_key, state)
 
     {current, pending_count, events, flush?} =
-      split_counting(batch_key, events, pending_count, false, current, partition_by)
+      split_counting(
+        batch_key,
+        events,
+        pending_count,
+        batch_splitter,
+        false,
+        current,
+        partition_by
+      )
+
+    acc =
+      deliver_or_update_batch(
+        batch_key,
+        current,
+        pending_count,
+        batch_splitter,
+        flush?,
+        timer,
+        acc,
+        state
+      )
 
-    acc = deliver_or_update_batch(batch_key, current, pending_count, flush?, timer, acc, state)
     handle_events_per_batch_key(events, acc, state)
   end
 
-  defp split_counting(batch_key, events, count, flush?, acc, partition_by) do
-    with [event | events] when count > 0 <- events,
-         ^batch_key <- batch_key(event, partition_by) do
-      flush? = flush? or event.batch_mode == :flush
-      split_counting(batch_key, events, count - 1, flush?, [event | acc], partition_by)
-    else
-      _ ->
-        {acc, count, events, flush?}
+  defp split_counting(_batch_key, [], count, _batch_splitter, flush?, acc, _partition_by) do
+    {acc, count, [], flush?}
+  end
+
+  defp split_counting(
+         batch_key,
+         [event | remained] = events,
+         {_flag, count} = split_counter,
+         batch_splitter,
+         flush?,
+         acc,
+         partition_by
+       ) do
+    event_batch_key = batch_key(event, partition_by)
+    {next_flag, _next_count} = next_split_counter = batch_splitter.(event, count)
+
+    cond do
+      event_batch_key != batch_key ->
+        # Switch to a different batch key
+        {acc, split_counter, events, flush?}
+
+      next_flag == :emit ->
+        # Batch splitter indicates a full batch
+        {[event | acc], next_split_counter, remained, flush?}
```

Comment:
There is a bug in the current implementation because `next_split_counter` is keeping the `{:emit, _} | {:cont, _}` tuple. This means that, if you change the `batch_key`, it will keep the result of `{:emit, _}` of the previous splitter, causing it to emit an element. Instead, this clause should be something like this:

    ```suggestion
      next_flag == :emit ->
        # Batch splitter indicates a full batch
        {[event | acc], next_count, remained, true}
```

I think then the `deliver_or_update_batch(batch_key, current, pending_count, flush?, timer, acc, state)` implementation can be kept exactly as it was. :)

Reply:
I have reviewed the logic and added a test case "customized batch splitter - one batcher different batch keys".  It does not seems to have the issue.

```elixir
      event_batch_key != batch_key ->
        # Switch to a different batch key
        {acc, split_counter, events, flush?}

      next_flag == :emit ->
        # Batch splitter indicates a full batch
        {[event | acc], next_split_counter, remained, flush?}
```

When the batch key is changed, the `acc` and `split_counter` are kept.  Batch will be saved in `deliver_or_update_batch`.  When `:emit`, the `next_split_counter` is for current `event` which is having the same batch key.